bookwyrm/models/bookwyrm_import_job.py

```diff
@@ -0,0 +1,515 @@
+            get_or_create_statuses(
+                job.user, models.Quotation, data.get("quotes"), book.id
+            )
+        if "include_lists" in job.required:
+            upsert_lists(job.user, data.get("lists"), data.get("list_items"), book.id)
+
+
+def get_or_create_edition(book_data, tar):
+    """Take a JSON string of book and edition data,
+    find or create the edition in the database and
+    return an edition instance"""
+
+    cover_path = book_data.get(
+        "cover", None
+    )  # we use this further down but need to assign a var before cleaning
+
+    clean_book = clean_values(book_data)
+    book = clean_book.copy()  # don't mutate the original book data
+
+    # prefer edition values only if they are not null
+    edition = clean_values(book["edition"])
+    for key in edition.keys():
+        if key not in book.keys() or (
+            key in book.keys() and (edition[key] not in [None, ""])
+        ):
+            book[key] = edition[key]
+
+    existing = find_existing(models.Edition, book)
+    if existing:
+        return existing
+
+    # the book is not in the local database, so we have to do this the hard way
+    local_authors = get_or_create_authors(book["authors"])
+
+    # get rid of everything that's not strictly in a Book
+    # or is many-to-many so can't be set directly
+    associated_values = [
+        "edition",
+        "authors",
+        "readthroughs",
+        "shelves",
+        "shelf_books",
+        "lists",
+        "list_items",
+        "reviews",
+        "comments",
+        "quotes",
+    ]
+
+    for val in associated_values:
+        del book[val]
+
+    # now we can save the book as an Edition
+    new_book = models.Edition.objects.create(**book)
+    new_book.authors.set(local_authors)  # now we can add authors with set()
+
+    # get cover from original book_data because we lost it in clean_values
+    if cover_path:
+        tar.write_image_to_file(cover_path, new_book.cover)
+
+    # NOTE: clean_values removes "last_edited_by"
+    # because it's a user ID from the old database
+    # if this is required, bookwyrm_export_job will
+    # need to bring in the user who edited it.
+
+    # create parent
+    work = models.Work.objects.create(title=book["title"])
+    work.authors.set(local_authors)
+    new_book.parent_work = work
+
+    new_book.save(broadcast=False)
+    return new_book
+
+
+def clean_values(data):
+    """clean values we don't want when creating new instances"""
+
+    values = [
+        "id",
+        "pk",
+        "remote_id",
+        "cover",
+        "preview_image",
+        "last_edited_by",
+        "last_edited_by_id",
+        "user",
+        "book_list",
+        "shelf_book",
+        "parent_work_id",
+    ]
+
+    common = data.keys() & values
+    new_data = data
+    for val in common:
+        del new_data[val]
+    return new_data
+
+
+def find_existing(cls, data):
+    """Given a book or author, find any existing model instances"""
```

Comment:
Happy to take advice on this and make things cleaner, but the primary reason we didn't use `model.find_existing` is that we absolutely do NOT want to match on ids. The incoming ids are likely from a different database, so they will match on a different model instance.

Reply:
It shouldn't ever match on id, since that's not a deduplication field. I wrote a couple test just to make sure I was right and I seem to be: https://github.com/bookwyrm-social/bookwyrm/pull/3084

Reply:
Ah, I see where I confused myself. I was thinking [this referred to the database id](https://github.com/bookwyrm-social/bookwyrm/blob/a93519ec3e45bf91dba96e48288e81fbc9bb3e18/bookwyrm/models/activitypub_mixin.py#L112C1-L114C54), but it's actually the ActivityPub id

Reply:
Ok @mouse-reeve I've tried this out. It's not going to work because `find_existing` from `activitypub_mixin` is looking for activitypub fieldnames, and we just have Bookwyrm class fieldnames.

This is only used for authors and editions, and those are relatively unproblematic to work with as ActivityPub JSON. So notwithstanding my comment that dealing with AP JSON was challenging, maybe we should take another look at how that might be done at least for book data?

Reply:
I think if it's possible to do with book data, that would have the most benefit, since there's pretty well tested code in place for converting an activitypub version of a book to a model entry. At minimum, this deduplicator function should use deduplication fields rather than enumerating them here

Reply:
ok ok ok @mouse-reeve this can work! _almost_.

It's totally broken right now but [see this PR in my repo](https://github.com/hughrun/bookwyrm/pull/1) so you can possibly find what's happening. Basically I now realise how we can pretty easily grab almost everything other than Readthroughs as AP JSON. I haven't looked at the import side yet but I'm reasonably confident about how it will be done.

One small problem: I'm getting an Exception from `DjangoJSONEncoder`:

```
bookwyrm-celery_worker-1   | User Export Job 37 Failed with error: Object of type _MISSING_TYPE is not JSON serializable
bookwyrm-celery_worker-1   | Traceback (most recent call last):
bookwyrm-celery_worker-1   |   File "/app/bookwyrm/models/bookwyrm_export_job.py", line 46, in start_export_task
bookwyrm-celery_worker-1   |     json_data = json_export(job.user)
bookwyrm-celery_worker-1   |   File "/app/bookwyrm/models/bookwyrm_export_job.py", line 214, in json_export
bookwyrm-celery_worker-1   |     return DjangoJSONEncoder().encode(data)
bookwyrm-celery_worker-1   |   File "/usr/local/lib/python3.9/json/encoder.py", line 199, in encode
bookwyrm-celery_worker-1   |     chunks = self.iterencode(o, _one_shot=True)
bookwyrm-celery_worker-1   |   File "/usr/local/lib/python3.9/json/encoder.py", line 257, in iterencode
bookwyrm-celery_worker-1   |     return _iterencode(o, 0)
bookwyrm-celery_worker-1   |   File "/usr/local/lib/python3.9/site-packages/django/core/serializers/json.py", line 105, in default
bookwyrm-celery_worker-1   |     return super().default(o)
bookwyrm-celery_worker-1   |   File "/usr/local/lib/python3.9/json/encoder.py", line 179, in default
bookwyrm-celery_worker-1   |     raise TypeError(f'Object of type {o.__class__.__name__} '
bookwyrm-celery_worker-1   | TypeError: Object of type _MISSING_TYPE is not JSON serializable
```

Looks like this is coming from the edition after it's exported via `to_activity()`:

```json
{
    "id": "https: //example.com/book/1",
    "type": "Edition",
    "links": <dataclasses._MISSING_TYPE object at 0xffff80d0faf0>,
    "fileLinks": [],
    "title": "Test book 1",
    "sortTitle": "Test book 1"
    ...
}
```

Not sure what's going on here, it feels like there's something wrong in `to_activity()` for `Link` fields.

Reply:
It's very weird that that's causing an error, but `links` _are_ in the activitypub spec for editions, even though they're not used, and I don't think are even in the model. Would removing it from `bookwyrm/activitypub/book.py` fix it?

Reply:
Aha! So they are. That explains it. The model has `file_links` but not `links`, so the `BookData` model is wrong (it has both). When I get home I'll try taking that out and see if it works.